aten/src/ATen/native/cuda/Gesv.cu

```diff
@@ -0,0 +1,133 @@
+#include "ATen/PinnedMemoryAllocator.h"
+#include "ATen/cuda/CUDAApplyUtils.cuh"
+
+#include "ATen/native/LinearAlgebraUtils.h"
+#include "ATen/native/Gesv.h"
+
+#include "THC.h" // for USE_MAGMA
+
+#ifdef USE_MAGMA
+#include <magma.h>
+#include <magma_types.h>
+#endif
+
+namespace at {
+namespace native {
+
+#ifdef USE_MAGMA
+template<class real>
+void magmaGesvBatched(
+    magma_int_t n, magma_int_t nrhs, real** dA_array, magma_int_t ldda,
+    magma_int_t** dipiv_array, real** dB_array, magma_int_t lddb,
+    magma_int_t* dinfo_array, magma_int_t batch_count, magma_queue_t queue) {
+  AT_ERROR("gesv only takes float or double Tensors");
+}
+
+template<>
+void magmaGesvBatched<float>(
+    magma_int_t n, magma_int_t nrhs, float** dA_array, magma_int_t ldda,
+    magma_int_t** dipiv_array, float** dB_array, magma_int_t lddb,
+    magma_int_t* dinfo_array, magma_int_t batch_count, magma_queue_t queue) {
+  magma_sgesv_batched(
+      n, nrhs, dA_array, ldda, dipiv_array,
+      dB_array, lddb, dinfo_array, batch_count, queue);
+}
+
+template<>
+void magmaGesvBatched<double>(
+    magma_int_t n, magma_int_t nrhs, double** dA_array, magma_int_t ldda,
+    magma_int_t** dipiv_array, double** dB_array, magma_int_t lddb,
+    magma_int_t* dinfo_array, magma_int_t batch_count, magma_queue_t queue) {
+  magma_dgesv_batched(
+      n, nrhs, dA_array, ldda, dipiv_array,
+      dB_array, lddb, dinfo_array, batch_count, queue);
+}
+
+static magma_queue_t createMagmaQueue(const Tensor& tensor) {
+  auto& context = tensor.type().get_context();
+  magma_queue_t magma_queue;
+  magma_queue_create_from_cuda(
+      tensor.get_device(),
+      context.getCurrentCUDAStream(),
+      THCState_getCurrentBlasHandle(context.thc_state),
+      THCState_getCurrentSparseHandle(context.thc_state),
+      &magma_queue);
+  return magma_queue;
+}
+#endif
+
+// Creates an array of size elements of type T, backed by pinned memory
+// wrapped in a Storage
+template<class T>
+static inline std::unique_ptr<Storage> pin_memory(int64_t size, Tensor dummy) {
+  int64_t adjusted_size = size * sizeof(T);
+  auto allocator = std::unique_ptr<Allocator>(new PinnedMemoryAllocator());
+  auto& backend = dummy.type().toBackend(kCPU).toScalarType(kByte);
+  return backend.storageWithAllocator(adjusted_size, std::move(allocator));
+}
+
+#define ALLOCATE_ARRAY(name, type, size, dummy_tensor) \
+  auto storage_##name = pin_memory<type>(size, dummy_tensor); \
+  name = reinterpret_cast<type*>(storage_##name->data());
+
+template <typename real>
+static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
+#ifndef USE_MAGMA
+AT_ERROR("gesv: MAGMA library not found in "
+    "compilation. Please rebuild with MAGMA.");
+#else
+  real* A_data = static_cast<real*>(A.data_ptr());
+  real* b_data = static_cast<real*>(b.data_ptr());
+  auto A_mat_stride = matrixStride(A);
+  auto b_mat_stride = matrixStride(b);
+
+  magma_int_t batch_size = batchCount(A);
```

Comment:
How large is `magma_int_t`?

Reply:
32-bit

Reply:
Then let's add a check here. `batchCount(A)` can be anything in `int64_t`.